sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -395,12 +396,19 @@ class Analyzer(
         Project(projectList = expanded, p.child)
       // If the aggregate function argument contains Stars, expand it.
       case a: Aggregate if containsStar(a.aggregateExpressions) =>
-        val expanded = a.aggregateExpressions.flatMap {
-          case s: Star => s.expand(a.child, resolver)
-          case o if containsStar(o :: Nil) => expandStarExpression(o, a.child) :: Nil
-          case o => o :: Nil
-        }.map(_.asInstanceOf[NamedExpression])
-        a.copy(aggregateExpressions = expanded)
+        if (conf.groupByOrdinal && a.groupingExpressions.exists(IntegerIndex.unapply(_).nonEmpty)) {
+          failAnalysis(
+            "Group by position: star is not allowed to use in the select list " +
+              "when using ordinals in group by")
+        } else {
+          val expanded = a.aggregateExpressions.flatMap {
+            case s: Star => s.expand(a.child, resolver)
+            case u @ UnresolvedAlias(_: Star, _) => expandStarExpression(u.child, a.child) :: Nil
```

Comment:
After reading the code, `Project` still has a problem in star expansion:
```scala
      val structDf = testData2.select("a", "b").as("record")
      structDf.select(hash($"record.*"))
```
 
Sorry, the previous PR does not cover all the cases. Let me submit a separate PR to handle all the star expansion. After the fix, we will not issue a message like:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L444

Of course, if we want to limit the support of star expansion in group by, we can do it for sure.